File: src/main/java/info/juliocnsouza/puzzles/arrays/SmallestDifference.java
package info.juliocnsouza.puzzles.arrays;

import java.util.Arrays;

// time: O(nLog(n) + mLog(m))
// space: O(1)
public class SmallestDifference {

    public int[] execute(int[] arg1, int[] arg2){
        Arrays.sort(arg1);
        Arrays.sort(arg2);

        if(arg1[0] < arg2[0]){
            return getSmallestDiffPair(arg1, arg2, false);
        }else{
            return getSmallestDiffPair(arg2, arg1, true);
        }
    }

    private int[] getSmallestDiffPair(int[] firstArr, int[] secondArr, boolean reverse){
        int firstIdx = 0;
        int secondIndex = 0;
        int difference = Integer.MAX_VALUE;
        int pair[] = new int[2];
        while(firstIdx < firstArr.length && secondIndex < secondArr.length){

            final int first = firstArr[firstIdx];
            final int second = secondArr[secondIndex];
            int currentDiff = Math.abs(first - second);

            if (currentDiff < difference){
                difference = currentDiff;
                pair[0] = reverse? second : first;
                pair[1] = reverse? first : second;
                if(difference == 0){
                    break;
                }
            }
            if (first > second){
                secondIndex++;
            }else{
                firstIdx++;
            }
        }
        return pair;
    }
}

File: javascript/algorithm/smallest_difference.js
module.exports = class SmallestDifference {
    exec(firstArr, secondArr) {
        firstArr.sort((a, b) => a - b);
        secondArr.sort((a, b) => a - b);
        let difference = Number.MAX_SAFE_INTEGER;
        let firstIdx = 0;
        let secondIndex = 0;
        const pair = [0, 0];
        while (firstIdx < firstArr.length && secondIndex < secondArr.length) {
            const first = firstArr[firstIdx];
            const second = secondArr[secondIndex];
            const currentDiff = Math.abs(first - second);

            if (currentDiff < difference) {
                difference = currentDiff;
                pair[0] = first;
                pair[1] = second;
                if (difference == 0) {
                    break;
                }
            }
            if (first > second) {
                secondIndex++;
            } else {
                firstIdx++;
            }
        }
        return pair;
    }

    dummyTest() {
        const arg1 = [-1, 5, 10, 28, 3];
        const arg2 = [26, 134, 135, 15, 17];
        const expected = [28, 26];
        const result = this.exec(arg1, arg2);
        console.log(JSON.stringify(result) == JSON.stringify(expected));
    }
};
